h2o-algos/src/main/java/hex/tree/GuidedSplitPoints.java

```diff
@@ -0,0 +1,159 @@
+ *   bins as we get deeper in the tree).
+ */
+public class GuidedSplitPoints {
+
+    static final double LOW_DENSITY_THRESHOLD = 0.2;
+
+    static boolean isApplicableTo(DHistogram h) {
+        return h._vals != null &&  // observations were not yet binned, we don't have the data to guide the splitting
+                h._isInt != 2 &&   // categorical columns have a specific handing
+                !h._intOpt;        // integer optimized columns have a single value per bin, no point in refining such bins
+    }
+
+    static double[] makeSplitPoints(DHistogram h, final int targetNBins, final double min, final double maxEx) {
+        // Collect bins to consider for refining
+        final List<BinDescriptor> bins = extractNonEmptyBins(h);
+
+        // Budget is given by target number of bins in the new layer, we keep all non-empty bins
+        final int totalBudget = targetNBins - bins.size() - 2; // how many bins we have to allocate (save 2 spots for min/max)
+        if (bins.isEmpty() || totalBudget <= 0)
+            return null;
+
+        int budgetLeft = totalBudget; // how many bins do we have left to redistribute
+
+        double totalSE = 0;
+        for (BinDescriptor bin : bins) {
+            totalSE += bin._se;
+        }
+
+        // For each bin find out how many new bins we can split it into
+        int[] newBinCounts = new int[bins.size()];
+        Collections.sort(bins); // sort by SE descending
+        for (int b = 0; budgetLeft > 0 && b < newBinCounts.length; b++) {
+            BinDescriptor bin = bins.get(b);
+            // distributed budget proportionally to SE
+            int newBins = Math.min((int) Math.ceil(totalBudget * bin._se / totalSE), budgetLeft);
+            budgetLeft -= newBins;
+            newBinCounts[b] = newBins;
+        }
+
+        // Define new split-points
+        final double[] customSplitPoints = new double[targetNBins - budgetLeft];
+        int i = 0;
+        for (int b = 0; b < newBinCounts.length; b++) {
+            BinDescriptor bin = bins.get(b);
+            customSplitPoints[i++] = bin._start;
+            double stepSize = (bin._end - bin._start) / (1 + newBinCounts[b]);
+            for (int s = 0; s < newBinCounts[b]; s++) {
+                customSplitPoints[i] = customSplitPoints[i - 1] + stepSize;
+                i++;
+            }
+        }
+        customSplitPoints[i++] = min; // This is based on QuantilesGlobal - DHistogram has assumption min/max will be in the split-points
+        customSplitPoints[i++] = h._maxIn;
+        assert i == customSplitPoints.length;
+
+        Arrays.sort(customSplitPoints);
+        return ArrayUtils.makeUniqueAndLimitToRange(customSplitPoints, min, maxEx);
+    }
+
+    static List<BinDescriptor> extractNonEmptyBins(DHistogram h) {
+        final int nonEmptyBins = h.nonEmptyBins();
+        final List<BinDescriptor> bins = new ArrayList<>(nonEmptyBins);
```

Comment:
So the size of bins is nonEmptyBins + h.nbins() without bins with weight == 0? I am a little bit confused about why.

Reply:
Ok, it makes sense to me now. Thanks for the explanation.